lib/Models/RegionMapping.js

```diff
@@ -639,22 +645,95 @@ function addDescriptionAndProperties(regionMapping, regionIndices, regionImagery
 
+function calculateImageryLayerIntervals(timeColumn, currentTime, animateForwards) {
+    // Before and after are defined according to animation direction
+    var isBefore = animateForwards ? JulianDate.lessThan : JulianDate.greaterThan;
+    var isAfter = animateForwards ? JulianDate.greaterThan : JulianDate.lessThan;
+    // Loop over time intervals and generate a time interval for the current and next imagery layer
+
+    // Need to keep track of 3 times:
+    // - past: The closest transition time in the past (for current interval)
+    // - future1: The closest transition time in the future (for current & next intervals)
+    // - future2: The next closest transition time in the future (for next interval)
+    var past = null;
+    var future1 = null;
+    var future2 = null;
+
+    for (const interval of timeColumn.timeIntervals) {
+        for (const time of [interval.start, interval.stop]) {
+            if (isAfter(time, currentTime)) {
+                if (future1 === null || isBefore(time, future1)) {
+                    future1 = time;
+                } else if (isAfter(time, future1) && (future2 === null || isBefore(time, future2))) {
+                    // Stop must always be after start
+                    future2 = time;
+                }
+            } else if (past === null || isAfter(time, past)) {
+                past = time;
+            }
+        }
+    }
+    var intervalFromDatePair = (date1, date2) => JulianDate.lessThan(date1, date2) ? new TimeInterval({start: date1, stop: date2}) : new TimeInterval({start: date2, stop: date1});
+    return {
+        currentInterval: (past && future1) ? intervalFromDatePair(past, future1) : null,
+        nextInterval: (future1 && future2) ? intervalFromDatePair(future1, future2) : null,
+    };
+}
+
 /**
  * Creates and enables a new ImageryLayer onto terria, showing appropriately colored regions.
  * @private
  * @param {RegionMapping} regionMapping    The table data source.
  * @param {Number} [layerIndex] The layer index of the new imagery layer.
+ *
  * @param {Array} [regionIndices] An array the same length as regionProvider.regions, mapping each region into the relevant index into the table data source.
  *                  If not provided, it is calculated, and failed/ambiguous warnings are displayed to the user.
  */
 function setNewRegionImageryLayer(regionMapping, layerIndex, regionIndices) {
-    regionMapping._imageryLayer = createNewRegionImageryLayer(regionMapping, layerIndex, regionIndices);
+    var catalogItem = regionMapping._catalogItem;
+    var currentTime = catalogItem.terria.clock.currentTime;
+    if (regionMapping._nextImageryLayerInterval && TimeInterval.contains(regionMapping._nextImageryLayerInterval, currentTime)) {
+        // Potential race condition. If someone changes variables after the last clock tick of the current imagery,
+        //  the old nextImageryLayer will be shown. Could store _nextILRegionIndices and compare before using that IL
```

Comment:
How big a problem is this?  Couldn't you fix it by setting `_nextImageryLayerInterval` to undefined when the variable changes?

Reply:
I think I may have fixed this race condition elsewhere, but I'm not sure